config/cuda.m4

```diff
@@ -0,0 +1,169 @@
+  fi
+  
+  AC_ARG_WITH([gpu_archs],
+              [AS_HELP_STRING([--with-gpu-archs=...],
+                              [default GPU architectures (default=detect)])],
+              [],
+              [with_gpu_archs='auto'])
+  if test "$HAVE_CUDA" = "1"; then
+    if test "$with_gpu_archs" = "auto"; then
+      AC_MSG_CHECKING([which CUDA architectures to target])
+
+      CXXFLAGS_save="$CXXFLAGS"
+      LDFLAGS_save="$LDFLAGS"
+      LIBS_save="$LIBS"
+      
+      LDFLAGS="-L$CUDA_HOME/lib64 -L$CUDA_HOME/lib"
+      LIBS="-lcuda -lcudart"
+      ac_run='$NVCC -o conftest$ac_ext $LDFLAGS $LIBS conftest.$ac_ext>&5'
+      AC_RUN_IFELSE([
+        AC_LANG_PROGRAM([[
+            #include <cuda.h>
+            #include <cuda_runtime.h>
+            #include <iostream>
+            #include <fstream>
+            #include <set>]],
+            [[
+            std::set<int> archs;
+            int major, minor, arch;
+            int deviceCount = 0;
+            cudaGetDeviceCount(&deviceCount);
+            if( deviceCount == 0 ) {
+              return 1;
+            }
+            std::ofstream fh;
+            fh.open("confarchs.out");
+            for(int dev=0; dev<deviceCount; dev++) {
+              cudaSetDevice(dev);
+              cudaDeviceGetAttribute(&major, cudaDevAttrComputeCapabilityMajor, dev);
+              cudaDeviceGetAttribute(&minor, cudaDevAttrComputeCapabilityMinor, dev);
+              arch = 10*major;
+              if( archs.count(arch) == 0 ) {
+                archs.insert(arch);
+                if( dev > 0 ) {
+                  fh << " ";
+                }
+                fh << arch;
+              }
+              arch += minor;
+              if( archs.count(arch) == 0 ) {
+                archs.insert(arch);
+                fh << " " << arch;
+              }
+            }
+            fh.close();]])],
+            [AC_SUBST([GPU_ARCHS], [`cat confarchs.out`])
+             ar_supported=$( ${NVCC} -h | ${GREP} -Po "'compute_[[0-9]]{2,3}" | cut -d_ -f2 | sort | uniq )
+             ar_valid=$( echo $GPU_ARCHS $ar_supported | xargs -n1 | sort | uniq -d | xargs )
+             AC_SUBST([GPU_ARCHS], [$ar_valid])
+             AC_MSG_RESULT([$GPU_ARCHS])],
+            [AC_MSG_ERROR(failed to find any)])
+
+      CXXFLAGS="$CXXFLAGS_save"
+      LDFLAGS="$LDFLAGS_save"
+      LIBS="$LIBS_save"
+    else
+      AC_SUBST([GPU_ARCHS], [$with_gpu_archs])
+    fi
+    
+    AC_MSG_CHECKING([for valid CUDA architectures])
+    ar_requested=$( echo "$GPU_ARCHS" | wc -w )
+    ar_supported=$( ${NVCC} -h | ${GREP} -Po "'compute_[[0-9]]{2,3}" | cut -d_ -f2 | sort | uniq )
+    ar_valid=$( echo $GPU_ARCHS $ar_supported | xargs -n1 | sort | uniq -d | xargs )
+    ar_found=$( echo $ar_valid | wc -w )
+    if test "$ar_requested" = "$ar_found"; then
+      AC_MSG_RESULT([yes])
+    else
+      AC_MSG_ERROR(only architectures $ar_valid are supported)
```

Comment:
I think this might need to be "only architectures $ar_supported" are supported.

I have occasionally (when specifying `--with-gpu-archs` incorrectly) seen the message come out as "only architectures  are supported". It looks like `$ar_supported` is what it gets from `nvcc -h`, and `$ar_valid` is the intersection of that with `$ar_requested` ?

Reply:
Yes, `$ar_supported` is everything supported by CUDA and `$ar_valid` is that is supported and what is requested.  I _think_ my original intention here was for this to catch and report when a partially valid list of archs. is provided.  So requesting something like "30 60" would tell you that only 60 was supported.  I hadn't thought about the case where nothing that was requested is supported.

Reply:
Maybe a better message would be "only '$ar_valid' of the requested archs. are supported"?

Reply:
Sure… when things go awry, it can be useful to have both, though it may not fit as nicely into the one-liner format configure likes:
```
Doing this... result is that
```
But usually `config.log` contains necessary detail. In the cases where I got a blank output there, I think it was because the test program failed to link... that error didn't appear on stdout, but it was in the log.

Reply:
[54ed45b](https://github.com/ledatelescope/bifrost/pull/157/commits/54ed45bfc809f35c74f008d468d9ef39be31fbe2) tries to do both.